--- dmenu.c_	2010-02-22 19:22:43.000000000 +0300
+++ dmenu.c	2010-02-23 00:14:56.000000000 +0300
@@ -255,9 +255,9 @@ drawtext(const char *text, unsigned long
 	if(fontxft) {
 		if (!dc.xftdraw)
 			eprint("error, creating xft drawable failed");
-		XftDrawString8(dc.xftdraw, &dc.xftcolor, dc.xftfont.xft_font, x, y, (unsigned char*)buf, len);
+		XftDrawStringUtf8(dc.xftdraw, &dc.xftcolor, dc.xftfont.xft_font, x, y, (unsigned char*)buf, len);
 	} else if(dc.font.set)
-		XmbDrawString(dpy, dc.drawable, dc.font.set, dc.gc, x, y, buf, len);
+		Xutf8DrawString(dpy, dc.drawable, dc.font.set, dc.gc, x, y, buf, len);
 	else
 		XDrawString(dpy, dc.drawable, dc.gc, x, y, buf, len);
 }
@@ -339,6 +339,14 @@ initxft() {
 	dc.xftfont.height = dc.xftfont.ascent + dc.xftfont.descent;
 }
 
+int
+firstutf8byte(unsigned char x) {
+	return ((x < 0x80) || 
+			((x & 0xE0) == 0xC0) || 
+			((x & 0xF0) == 0xE0) ||
+			((x & 0xF8) == 0xF0));
+}
+
 void
 kpress(XKeyEvent * e) {
 	char buf[32];
@@ -432,7 +440,10 @@ kpress(XKeyEvent * e) {
 		break;
 	case XK_BackSpace:
 		if(len) {
-			text[--len] = 0;
+			do {
+				len--;
+			} while ( ! firstutf8byte(text[len]));
+			text[len] = 0;
 			match(text);
 		}
 		break;
@@ -694,11 +705,11 @@ setup(Bool topbar) {
 int
 textnw(const char *text, unsigned int len) {
 	if(fontxft) {
-		XftTextExtents8(dpy, dc.xftfont.xft_font, (const FcChar8*)text, len, &dc.gi);
+		XftTextExtentsUtf8(dpy, dc.xftfont.xft_font, (const FcChar8*)text, len, &dc.gi);
 		return dc.gi.width;
     } else if(dc.font.set) {
 	    XRectangle r;
-		XmbTextExtents(dc.font.set, text, len, NULL, &r);
+		Xutf8TextExtents(dc.font.set, text, len, NULL, &r);
 		return r.width;
 	}
 	return XTextWidth(dc.font.xfont, text, len);
