diff --git a/Swiften/Serializer/StreamResumeSerializer.cpp b/Swiften/Serializer/StreamResumeSerializer.cpp
index e9e520d..b576c78 100644
--- a/Swiften/Serializer/StreamResumeSerializer.cpp
+++ b/Swiften/Serializer/StreamResumeSerializer.cpp
@@ -8,6 +8,7 @@
 
 #include <boost/shared_ptr.hpp>
 #include <boost/lexical_cast.hpp>
+#include <boost/optional/optional_io.hpp>
 
 #include <Swiften/Elements/StreamResume.h>
 #include <Swiften/Serializer/XML/XMLElement.h>
diff --git a/Swiften/Serializer/StreamResumedSerializer.cpp b/Swiften/Serializer/StreamResumedSerializer.cpp
index 7ae82d1..2f98fc9 100644
--- a/Swiften/Serializer/StreamResumedSerializer.cpp
+++ b/Swiften/Serializer/StreamResumedSerializer.cpp
@@ -8,6 +8,7 @@
 
 #include <boost/shared_ptr.hpp>
 #include <boost/lexical_cast.hpp>
+#include <boost/optional/optional_io.hpp>
 
 #include <Swiften/Elements/StreamResumed.h>
 #include <Swiften/Serializer/XML/XMLElement.h>
