--- onscripter-en-20080214/configure.old	2008-02-13 00:02:19.000000000 +0100
+++ onscripter-en-20080214/configure	2008-02-22 15:21:43.031293904 +0100
@@ -743,7 +743,7 @@
 
 LIBS = -Lextlib/lib \\
        $LINKSDL_image \$(if \$(findstring true,$INTERNAL_SDL_IMAGE),$LINKjpeg $LINKpng $LINKz) \\
-       $LINKSDL_mixer \$(if \$(findstring true,$INTERNAL_SDL_MIXER),$LINKvorbisfile $LINKvorbis $LINKogg) \\
+       $LINKSDL_mixer $LINKvorbisfile \$(if \$(findstring true,$INTERNAL_SDL_MIXER),$LINKvorbis $LINKogg) \\
        \$(shell $SDL_CONFIG --libs)      \\
        \$(shell $SMPEG_CONFIG --libs)    \\
        $LINKSDL_ttf \$(shell $FREETYPE_CONFIG --libs) \\
@@ -763,7 +763,7 @@
 .SUFFIXES:
 .SUFFIXES: $OBJ .cpp .h .c
 
-OSCFLAGS = -O2 -Wpointer-arith -Werror -fomit-frame-pointer -pipe \$(OSCFLAGSEXTRA)
+OSCFLAGS = \$(CXXFLAGS) \$(OSCFLAGSEXTRA)
 
 TARGET = onscripter$EXE
 
