--- trunk/KDE/kdelibs/khtml/html/html_objectimpl.cpp	2007/12/27 04:54:15	753267
+++ trunk/KDE/kdelibs/khtml/html/html_objectimpl.cpp	2007/12/27 04:54:22	753268
@@ -243,7 +243,12 @@
 
     _style->deref();
     NodeBaseImpl::attach();
+    
+    if (m_render) QTimer::singleShot( 0, this, SLOT( slotEmitLoadEvent() ) );
+}
 
+void HTMLObjectBaseElementImpl::slotEmitLoadEvent()
+{
     // ### do this when we are actually finished loading instead
     if (m_render) dispatchHTMLEvent(EventImpl::LOAD_EVENT, false, false);
 }
--- trunk/KDE/kdelibs/khtml/html/html_objectimpl.h	2007/12/27 04:54:15	753267
+++ trunk/KDE/kdelibs/khtml/html/html_objectimpl.h	2007/12/27 04:54:22	753268
@@ -62,6 +62,7 @@
     virtual void removeId(const QString& id);
 protected Q_SLOTS:
     void slotRenderAlternative();
+    void slotEmitLoadEvent();
 protected:
     DOMString     m_name;
 };
