--- golded3/golded.h.orig	2007-01-08 12:48:18.000000000 +0300
+++ golded3/golded.h	2008-09-13 22:54:55.000000000 +0400
@@ -27,6 +27,9 @@
 #ifndef __GOLDED_H
 #define __GOLDED_H
 
+#ifndef INT_MAX
+#define INT_MAX 214783647
+#endif
 
 //  ------------------------------------------------------------------
 //  Standard library headers
--- goldlib/gall/gdbgerr.cpp.orig	2000-02-25 13:11:57.000000000 +0300
+++ goldlib/gall/gdbgerr.cpp	2008-09-13 22:50:06.000000000 +0400
@@ -28,6 +28,7 @@
 #include <gutlmisc.h>
 #include <gfilutil.h>
 #include <gdbgerr.h>
+#include <string.h>
 
 
 //  ------------------------------------------------------------------
--- goldlib/gall/gdbgtrk.cpp.orig	2005-10-07 16:41:11.000000000 +0400
+++ goldlib/gall/gdbgtrk.cpp	2008-09-13 22:52:43.000000000 +0400
@@ -28,7 +28,8 @@
 #include <gdbgtrk.h>
 #include <gmemdbg.h>
 #include <glog.h>
-
+#include <malloc.h>
+#include <stdlib.h>
 
 //  ------------------------------------------------------------------
 
--- goldlib/gall/gespell.h.orig	2006-11-25 14:44:40.000000000 +0300
+++ goldlib/gall/gespell.h	2008-09-13 22:55:35.000000000 +0400
@@ -27,6 +27,9 @@
 #ifndef __GESPELL_H__
 #define __GESPELL_H__
 
+#ifndef PATH_MAX
+#define PATH_MAX 256
+#endif
 
 //  ------------------------------------------------------------------
 
--- goldlib/gall/gfilutl1.cpp.orig	2006-04-26 21:06:23.000000000 +0400
+++ goldlib/gall/gfilutl1.cpp	2008-09-13 22:56:07.000000000 +0400
@@ -29,6 +29,7 @@
 #include <gstrall.h>
 #include <gfilutil.h>
 #include <gfile.h>
+#include <stdlib.h>
 
 #if defined(__MINGW32__) || defined(_MSC_VER)
 #include <sys/utime.h>
--- goldlib/gall/gfilutl2.cpp.orig	2006-01-15 00:06:37.000000000 +0300
+++ goldlib/gall/gfilutl2.cpp	2008-09-13 22:56:36.000000000 +0400
@@ -26,6 +26,8 @@
 //  ------------------------------------------------------------------
 
 #include <gfilutil.h>
+#include <stdlib.h>
+#include <string.h>
 #ifndef __HAVE_DRIVES__
 #include <pwd.h>
 #endif
--- goldlib/gall/gftnnl.cpp.orig	2000-02-25 13:11:01.000000000 +0300
+++ goldlib/gall/gftnnl.cpp	2008-09-13 22:57:04.000000000 +0400
@@ -27,7 +27,7 @@
 #include <cstdio>
 #include <gstrall.h>
 #include <gftnnl.h>
-
+#include <stdlib.h>
 
 //  ------------------------------------------------------------------
 
--- goldlib/gall/gftnnlfd.cpp.orig	2000-02-25 13:11:02.000000000 +0300
+++ goldlib/gall/gftnnlfd.cpp	2008-09-13 22:57:29.000000000 +0400
@@ -29,7 +29,7 @@
 #include <gfilutil.h>
 #include <gstrall.h>
 #include <gftnnlfd.h>
-
+#include <stdlib.h>
 
 //  ------------------------------------------------------------------
 
--- goldlib/gall/gftnnlge.cpp.orig	2005-11-01 01:18:26.000000000 +0300
+++ goldlib/gall/gftnnlge.cpp	2008-09-13 22:57:56.000000000 +0400
@@ -30,7 +30,7 @@
 #include <gmemdbg.h>
 #include <gstrall.h>
 #include <gftnnlge.h>
-
+#include <stdlib.h>
 
 //  ------------------------------------------------------------------
 
--- goldlib/gall/gmemdbg.cpp.orig	2008-03-01 07:24:49.000000000 +0300
+++ goldlib/gall/gmemdbg.cpp	2008-09-13 22:58:35.000000000 +0400
@@ -31,7 +31,9 @@
 #include <gmemdbg.h>
 #include <glog.h>
 #include <gmemall.h>
-
+#include <malloc.h>
+#include <stdlib.h>
+#include <string.h>
 
 //  ------------------------------------------------------------------
 //  Global vars
--- goldlib/gall/gstrutil.cpp.orig	2007-01-19 20:38:41.000000000 +0300
+++ goldlib/gall/gstrutil.cpp	2008-09-13 22:59:05.000000000 +0400
@@ -30,6 +30,8 @@
 #include <gstrall.h>
 #include <glog.h>
 #include <gdbgerr.h>
+#include <stdlib.h>
+#include <string.h>
 
 extern glog LOG;
 
--- goldlib/gall/gutlgrp.cpp.orig	2002-09-24 14:16:03.000000000 +0400
+++ goldlib/gall/gutlgrp.cpp	2008-09-13 22:59:38.000000000 +0400
@@ -33,7 +33,7 @@
 #include <string>
 #include <gwildmat.h>
 #include <gutlgrp.h>
-
+#include <string.h>
 
 //  ------------------------------------------------------------------
 //  Constructor
--- rddt/rddt.cpp.orig	2007-01-08 15:09:42.000000000 +0300
+++ rddt/rddt.cpp	2008-09-13 23:00:09.000000000 +0400
@@ -38,6 +38,7 @@
 #include <gftnall.h>
 #include <golded3.h>
 #include <gmemdbg.h>
+#include <stdlib.h>
 
 int debug = false;
 
--- golded3/geline.cpp.orig	2008-01-20 21:02:59.000000000 +0300
+++ golded3/geline.cpp	2008-09-14 08:41:51.000000000 +0400
@@ -29,7 +29,7 @@
 #include <gstrmail.h>
 #include <gutlcode.h>
 #include <ghdrmime.h>
-
+#include <limits.h>
 #if defined(__USE_ALLOCA__)
   #include <malloc.h>
 #endif
--- golded3/gealst.cpp.orig	2007-01-08 15:09:41.000000000 +0300
+++ golded3/gealst.cpp	2008-09-14 08:40:31.000000000 +0400
@@ -32,7 +32,7 @@
 #endif
 #include <algorithm>
 #include <golded.h>
-
+#include <limits.h>
 #if defined(__USE_ALLOCA__)
 #include <malloc.h>
 #endif
