--- Config.def.orig	2007-02-24 17:23:03.000000000 +0300
+++ Config.def	2008-09-14 01:04:06.000000000 +0400
@@ -76,9 +76,9 @@
 # Uncomment to disable Synchronet BBS support
 #CPPFLAGS+=-DGCFG_NOSYNCHRONET
 # Uncomment to disable MS Office spellchecker support (win32 only)
-#CPPFLAGS+=-DGCFG_NO_MSSPELL
+CPPFLAGS+=-DGCFG_NO_MSSPELL
 # Uncomment to disable HunSpell spellchecker support (multiplatform)
-#CPPFLAGS+=-DGCFG_NO_MYSPELL
+CPPFLAGS+=-DGCFG_NO_MYSPELL
 # Uncomment to disable regex search feature
 #CPPFLAGS+=-DGCFG_NO_REGEX
 # Comment to disable AreaRecycleBin keyword and feature
