diff --git a/dlls/wined3d/device.c b/dlls/wined3d/device.c
index 212184c..abcd1b7 100644
--- a/dlls/wined3d/device.c
+++ b/dlls/wined3d/device.c
@@ -3304,6 +3304,8 @@ static HRESULT WINAPI IWineD3DDeviceImpl_SetSamplerState(IWineD3DDevice *iface,
     TRACE("(%p) : Sampler %#x, Type %s (%#x), Value %#x\n",
             This, Sampler, debug_d3dsamplerstate(Type), Type, Value);
 
+    if (Type == WINED3DSAMP_SRGBTEXTURE) return WINED3D_OK;
+
     if (Sampler >= WINED3DVERTEXTEXTURESAMPLER0 && Sampler <= WINED3DVERTEXTEXTURESAMPLER3) {
         Sampler -= (WINED3DVERTEXTEXTURESAMPLER0 - MAX_FRAGMENT_SAMPLERS);
     }
