--- makefile.unix	2010-10-12 00:54:46.000000000 +0400
+++ makefile.unix	2010-10-14 08:22:19.969181833 +0400
@@ -3,36 +3,27 @@
 # file license.txt or http://www.opensource.org/licenses/mit-license.php.
 
 
-INCLUDEPATHS= \
- -I"/usr/local/include/wx-2.9" \
- -I"/usr/local/lib/wx/include/gtk2-unicode-debug-static-2.9"
+INCLUDEPATHS=-I /usr/include/db4.8/
 
 # for wxWidgets 2.9.1, add -l Xxf86vm
-WXLIBS= \
- -Wl,-Bstatic \
-   -l wx_gtk2ud-2.9 \
- -Wl,-Bdynamic \
-   -l gtk-x11-2.0 \
-   -l SM
+WXLIBS=$(shell wx-config-2.9 --libs)
 
 # for boost 1.37, add -mt to the boost libraries
 LIBS= \
- -Wl,-Bstatic \
    -l boost_system \
    -l boost_filesystem \
    -l boost_program_options \
    -l boost_thread \
    -l db_cxx \
    -l ssl \
-   -l crypto \
- -Wl,-Bdynamic \
-   -l gthread-2.0 \
+   $(shell pkg-config --libs libcrypto) \
+   $(shell pkg-config --libs gthread-2.0) \
    -l z \
    -l dl
 
-DEFS=-D__WXGTK__ -DNOPCH -DFOURWAYSSE2 -DUSE_SSL
+DEFS=$(shell wx-config-2.9 --cxxflags) -DFOURWAYSSE2 -DUSE_SSL
 DEBUGFLAGS=-g -D__WXDEBUG__
-CFLAGS=-O2 -Wno-invalid-offsetof -Wformat $(DEBUGFLAGS) $(DEFS) $(INCLUDEPATHS)
+CFLAGS=$(CXXFLAGS) -Wno-invalid-offsetof -Wformat $(DEFS) $(INCLUDEPATHS)
 HEADERS=headers.h strlcpy.h serialize.h uint256.h util.h key.h bignum.h base58.h \
     script.h db.h net.h irc.h main.h rpc.h uibase.h ui.h noui.h init.h
 
