--- a/src/makefile.unix	2010-06-25 04:02:00.000000000 +0400
+++ b/src/makefile.unix	2010-07-17 14:22:15.000000000 +0400
@@ -4,31 +4,19 @@
 
 
 
-INCLUDEPATHS= \
- -I"/usr/include" \
- -I"/usr/local/include/wx-2.9" \
- -I"/usr/local/lib/wx/include/gtk2-unicode-debug-static-2.9"
-
-LIBPATHS= \
- -L"/usr/lib" \
- -L"/usr/local/lib"
-
-WXLIBS= \
- -Wl,-Bstatic \
-   -l wx_gtk2ud-2.9 \
- -Wl,-Bdynamic \
-   -l gtk-x11-2.0 -l SM
+INCLUDEPATHS=-I /usr/include/db4.8/
+
+LIBPATHS=-L"/usr/lib"
+
+WXLIBS=$(shell wx-config-2.9 --libs)
 
 LIBS= \
- -Wl,-Bstatic \
    -l boost_system -l boost_filesystem \
    -l db_cxx \
-   -l crypto \
- -Wl,-Bdynamic \
-   -l gthread-2.0
+   $(shell pkg-config --libs libcrypto) \
+   $(shell pkg-config --libs gthread-2.0)
 
-WXDEFS=-D__WXGTK__ -DNOPCH
-DEBUGFLAGS=-g -D__WXDEBUG__
+WXDEFS=$(shell wx-config-2.9 --cxxflags)
 CFLAGS=-O2 -Wno-invalid-offsetof -Wformat $(DEBUGFLAGS) $(WXDEFS) $(INCLUDEPATHS)
 HEADERS=headers.h strlcpy.h serialize.h uint256.h util.h key.h bignum.h base58.h \
     script.h db.h net.h irc.h main.h rpc.h uibase.h ui.h init.h sha.h
