diff --git a/Makefile b/Makefile
index ea451fe..5a9b6cd 100644
--- a/Makefile
+++ b/Makefile
@@ -3,7 +3,6 @@ include config.mk
 all:
 	$(MAKE) -C icebox
 	$(MAKE) -C icepack
-	$(MAKE) -C iceprog
 	$(MAKE) -C icemulti
 	$(MAKE) -C icepll
 	# $(MAKE) -C icetime
@@ -11,7 +10,6 @@ all:
 clean:
 	$(MAKE) -C icebox clean
 	$(MAKE) -C icepack clean
-	$(MAKE) -C iceprog clean
 	$(MAKE) -C icemulti clean
 	$(MAKE) -C icepll clean
 	# $(MAKE) -C icetime clean
@@ -19,7 +17,6 @@ clean:
 install:
 	$(MAKE) -C icebox install
 	$(MAKE) -C icepack install
-	$(MAKE) -C iceprog install
 	$(MAKE) -C icemulti install
 	$(MAKE) -C icepll install
 	# $(MAKE) -C icetime install
@@ -27,7 +24,6 @@ install:
 uninstall:
 	$(MAKE) -C icebox uninstall
 	$(MAKE) -C icepack uninstall
-	$(MAKE) -C iceprog uninstall
 	$(MAKE) -C icemulti uninstall
 	$(MAKE) -C icepll uninstall
 	# $(MAKE) -C icetime uninstall
