diff -urN a.orig/src/groupchatdlg.cpp a/src/groupchatdlg.cpp
--- a.orig/src/groupchatdlg.cpp	2009-03-21 19:44:06.000000000 +0300
+++ a/src/groupchatdlg.cpp	2009-03-21 21:04:38.000000000 +0300
@@ -1464,6 +1464,8 @@
 
 QString GCMainDlg::getNickColor(QString nick)
 {
+	unsigned char hash[4];
+	char color[8], *cnick;
 	int sender;
 	if(nick == d->self||nick.isEmpty())
 		sender = -1;
@@ -1478,15 +1480,19 @@
 	
 	QStringList nickColors = PsiOptions::instance()->getOption("options.ui.look.colors.muc.nick-colors").toStringList();
 	
-	if(!PsiOptions::instance()->getOption("options.ui.muc.use-nick-coloring").toBool() || nickColors.empty()) {
+	if(!PsiOptions::instance()->getOption("options.ui.muc.use-nick-coloring").toBool()) {
 		return "#000000";
 	}
-	else if(sender == -1 || nickColors.size() == 1) {
-		return nickColors[nickColors.size()-1];
-	}
 	else {
-		int n = sender % (nickColors.size()-1);
-		return nickColors[n];
+		unsigned nicklen, i, *h = reinterpret_cast<unsigned *>(hash);
+		cnick = const_cast<char *>(nick.ascii());
+		nicklen = strlen(cnick);
+		*h = 12345;
+		for (i = 0; i < nicklen; i++) {
+			*h *= cnick[i];
+		}
+		sprintf(color, "#%02hhX%02hhX%02hhX", hash[0], hash[1], hash[2]);
+		return QString(color);
 	}
 }
 
