diff -urNp psi-0.10.org/src/tools/openpgp/gpgop.cpp psi-0.10/src/tools/openpgp/gpgop.cpp
--- psi-0.10.org/src/tools/openpgp/gpgop.cpp	2005-08-21 20:44:28.000000000 +0300
+++ psi-0.10/src/tools/openpgp/gpgop.cpp	2007-06-28 18:55:55.000000000 +0300
@@ -377,7 +377,7 @@ void GpgOp::proc_statusLine(const QStrin
 		rest = str.mid(n+1);
 	}
 
-	if(s == "NEED_PASSPHRASE") {
+	if(s == "GET_HIDDEN") {
 		if(!(d->tryAgent && getenv("GPG_AGENT_INFO"))) {
 			if(!d->didPassphrase) {
 				d->didPassphrase = true;
