diff -ru psi-dev-snapshot-2007-02-04-orig/src/plugins/generic/echo/echoplugin.cpp psi-dev-snapshot-2007-02-04/src/plugins/generic/echo/echoplugin.cpp
--- psi-dev-snapshot-2007-02-04-orig/src/plugins/generic/echo/echoplugin.cpp	2007-02-04 06:30:25.000000000 +0100
+++ psi-dev-snapshot-2007-02-04/src/plugins/generic/echo/echoplugin.cpp	2007-02-05 02:29:04.000000000 +0100
@@ -60,7 +60,7 @@
 {
 	qWarning(qPrintable(QString("Received message '%1', echoing back to %2").arg(message).arg(fromDisplay)));
 	QVariant option;
-	emit getGlobalOption(message, &option);
+	emit getGlobalOption(message, option);
 	QString reply;
 	if (option.isValid())
 		reply=QString("<message to=\"%1\" type=\"chat\"><body>Option %2 = %3 </body></message>").arg(fromJid).arg(message).arg(option.toString());
