diff -Naur psi-dev-snapshot-2007-01-12-orig/src/pluginmanager.cpp psi-dev-snapshot-2007-01-12/src/pluginmanager.cpp
--- psi-dev-snapshot-2007-01-12-orig/src/pluginmanager.cpp	2007-01-12 06:30:08.000000000 +0100
+++ psi-dev-snapshot-2007-01-12/src/pluginmanager.cpp	2007-01-13 01:54:03.000000000 +0100
@@ -422,7 +422,6 @@
  */ 
 void PluginManager::sendStanza( const PsiAccount* account, const QString& stanza)
 {
-	qDebug(qPrintable(QString("Want to send stanza  to account %2").arg((int)account)));
 	if (!clients_.contains(account) || !verifyStanza(stanza))
 		return;
 	clients_[account]->send(stanza);
