diff -r 4a485db25506 configure.in
--- a/configure.in	Tue Mar 06 13:18:02 2007 +0100
+++ b/configure.in	Sat Mar 17 19:25:53 2007 +0200
@@ -844,6 +844,7 @@ fi
 fi
 AC_TRY_RUN([
 #include <stdio.h>
+#include <stdlib.h>
 #include "$KERNDIR/include/linux/autoconf.h"
 int main(void) {
 	FILE *file = fopen("processor.id", "w+");
@@ -861,7 +862,7 @@ int main(void) {
 	fprintf(file, "i586tsc");
 #elif defined(CONFIG_M586MMX)
 	fprintf(file, "i586mmx");
-#elif defined(CONFIG_M686) || defined(CONFIG_MPENTIUMII) || defined(CONFIG_MPENTIUMIII) || defined(CONFIG_MPENTIUM4) || defined(CONFIG_MPENTIUMM)
+#elif defined(CONFIG_M686) || defined(CONFIG_MPENTIUMII) || defined(CONFIG_MPENTIUMIII) || defined(CONFIG_MPENTIUM4) || defined(CONFIG_MPENTIUMM) || defined(CONFIG_MCORE2)
 	fprintf(file, "i686");
 #elif defined(CONFIG_MK6)
 	fprintf(file, "k6");
