Index: configure.ac
===================================================================
--- configure.ac	(revision 7810)
+++ configure.ac	(working copy)
@@ -56,10 +56,6 @@
 AC_PROG_CC
 AC_PROG_CXX
 AC_C_INLINE
-if test "x$GCC" = "xyes" ; then
-    CFLAGS="-std=gnu99 -ggdb3 -Wall -Wextra -Wredundant-decls -Wpointer-arith -Wformat-security -W -Wmissing-declarations -Wdeclaration-after-statement -Wcast-align -Winit-self -Wundef -Wnested-externs -Wmissing-format-attribute"
-    CXXFLAGS="-Wall -Wformat-security -W"
-fi
 
 AC_HEADER_STDC
 AC_HEADER_TIME
