--- Makefile	2015-06-25 02:18:04.000000000 +0300
+++ Makefile	2015-07-28 01:07:08.652063756 +0300
@@ -25,7 +25,7 @@
 CONFDIR=/etc/apparmor
 INSTALL_CONFDIR=${DESTDIR}${CONFDIR}
 LOCALEDIR=/usr/share/locale
-MANPAGES=apparmor.d.5 apparmor.7 apparmor_parser.8 subdomain.conf.5
+MANPAGES=apparmor.d.5 apparmor.7 apparmor_parser.8
 
 YACC	:= /usr/bin/bison
 YFLAGS	:= -d
@@ -370,11 +370,9 @@
 .PHONY: install-indep
 install-indep:
 	install -m 755 -d $(INSTALL_CONFDIR)
-	install -m 644 subdomain.conf $(INSTALL_CONFDIR)
+	install -m 755 -d ${DESTDIR}/usr/libexec
 	install -m 644 parser.conf $(INSTALL_CONFDIR)
-	install -m 755 -d ${DESTDIR}/var/lib/apparmor
-	install -m 755 -d $(APPARMOR_BIN_PREFIX)
-	install -m 755 rc.apparmor.functions $(APPARMOR_BIN_PREFIX)
+	install -m 755 rc.apparmor.functions ${DESTDIR}/usr/libexec
 	$(MAKE) -C po install NAME=${NAME} DESTDIR=${DESTDIR}
 	$(MAKE) install_manpages DESTDIR=${DESTDIR}
 
