diff --git a/src/emc/rs274ngc/interp_setup.cc b/src/emc/rs274ngc/interp_setup.cc
index 2120991..7355c6c 100644
--- a/src/emc/rs274ngc/interp_setup.cc
+++ b/src/emc/rs274ngc/interp_setup.cc
@@ -179,7 +179,7 @@ setup_struct::setup_struct() :
 {
 }
 
-setup::~setup() {
+setup_struct::~setup_struct() {
     assert(!pythis || Py_IsInitialized());
     if(pythis) delete pythis;
 }
diff --git a/src/hal/user_comps/mb2hal/mb2hal.c b/src/hal/user_comps/mb2hal/mb2hal.c
index 1ba9a87..d46a43d 100644
--- a/src/hal/user_comps/mb2hal/mb2hal.c
+++ b/src/hal/user_comps/mb2hal/mb2hal.c
@@ -307,7 +307,6 @@ retCode get_tx_connection(const int this_mb_tx_num, int *ret_connected)
     mb_tx_t   *this_mb_tx;
     mb_link_t *this_mb_link;
     int        this_mb_link_num;
-    struct timeval timeout;
 
     if (this_mb_tx_num < 0 || this_mb_tx_num > gbl.tot_mb_tx) {
         ERR(gbl.init_dbg, "parameter out of range this_mb_tx_num[%d]", this_mb_tx_num);
@@ -357,16 +356,16 @@ retCode get_tx_connection(const int this_mb_tx_num, int *ret_connected)
     modbus_set_debug(this_mb_link->modbus, this_mb_tx->protocol_debug);
 
     //set response and byte timeout according to each mb_tx
-    timeout.tv_sec  = this_mb_tx->mb_response_timeout_ms / 1000;
-    timeout.tv_usec = (this_mb_tx->mb_response_timeout_ms % 1000) * 1000;
-    modbus_set_response_timeout(this_mb_link->modbus, &timeout);
+    modbus_set_response_timeout(this_mb_link->modbus,
+            this_mb_tx->mb_response_timeout_ms / 1000,
+            (this_mb_tx->mb_response_timeout_ms % 1000) * 1000);
     //DBG(this_mb_tx->cfg_debug, "mb_tx_num[%d] mb_links[%d] response timeout [%d] ([%d] [%d])",
     //    this_mb_tx_num, this_mb_tx->mb_link_num, this_mb_tx->mb_response_timeout_ms,
     //    (int) timeout.tv_sec, (int) timeout.tv_usec);
 
-    timeout.tv_sec  = this_mb_tx->mb_byte_timeout_ms / 1000;
-    timeout.tv_usec = (this_mb_tx->mb_byte_timeout_ms % 1000) * 1000;
-    modbus_set_byte_timeout(this_mb_link->modbus, &timeout);
+    modbus_set_byte_timeout(this_mb_link->modbus,
+            this_mb_tx->mb_byte_timeout_ms / 1000,
+            (this_mb_tx->mb_byte_timeout_ms % 1000) * 1000);
     //DBG(this_mb_tx->cfg_debug, "mb_tx_num[%d] mb_links[%d] byte timeout [%d] ([%d] [%d])",
     //    this_mb_tx_num, this_mb_tx->mb_link_num, this_mb_tx->mb_byte_timeout_ms,
     //    (int) timeout.tv_sec, (int) timeout.tv_usec);
