diff -urN crda-1.0.1.orig/crda.c crda-1.0.1/crda.c
--- crda-1.0.1.orig/crda.c	2009-02-13 19:08:28.000000000 +0300
+++ crda-1.0.1/crda.c	2009-02-13 20:36:58.000000000 +0300
@@ -127,12 +127,12 @@
 	freq  = crda_get_file_ptr(db, dblen, sizeof(*freq), rule->freq_range_ptr);
 	power = crda_get_file_ptr(db, dblen, sizeof(*power), rule->power_rule_ptr);
 
-	NLA_PUT_U32(msg, NL80211_ATTR_REG_RULE_FLAGS,		ntohl(rule->flags));
-	NLA_PUT_U32(msg, NL80211_ATTR_FREQ_RANGE_START,		ntohl(freq->start_freq));
-	NLA_PUT_U32(msg, NL80211_ATTR_FREQ_RANGE_END,		ntohl(freq->end_freq));
-	NLA_PUT_U32(msg, NL80211_ATTR_FREQ_RANGE_MAX_BW,	ntohl(freq->max_bandwidth));
-	NLA_PUT_U32(msg, NL80211_ATTR_POWER_RULE_MAX_ANT_GAIN,	ntohl(power->max_antenna_gain));
-	NLA_PUT_U32(msg, NL80211_ATTR_POWER_RULE_MAX_EIRP,	ntohl(power->max_eirp));
+	NLA_PUT_U32(msg, NL80211_ATTR_REG_RULE_FLAGS,		0);
+	NLA_PUT_U32(msg, NL80211_ATTR_FREQ_RANGE_START,		1);
+	NLA_PUT_U32(msg, NL80211_ATTR_FREQ_RANGE_END,		7000000);
+	NLA_PUT_U32(msg, NL80211_ATTR_FREQ_RANGE_MAX_BW,	40000);
+	NLA_PUT_U32(msg, NL80211_ATTR_POWER_RULE_MAX_ANT_GAIN,	0);
+	NLA_PUT_U32(msg, NL80211_ATTR_POWER_RULE_MAX_EIRP,	10000);
 
 	return 0;
 
@@ -286,6 +286,7 @@
 			goto nla_put_failure;
 
 		nla_nest_end(msg, nl_reg_rule);
+		break;
 	}
 
 	nla_nest_end(msg, nl_reg_rules);
