--- configure.ac
+++ configure.ac
@@ -1250,7 +1250,7 @@
 
 
 dnl Check for xxd
-AC_CHECK_PROG(xxd_found, xxd, yes)
+AC_CHECK_PROG(xxd_found, xxdi.pl, yes)
 if test "$xxd_found" != yes ;then
  	AC_MSG_ERROR("xxd is required (provided by vim package)")
 fi
--- src/Makefile.am
+++ src/Makefile.am
@@ -471,11 +471,11 @@
 
 yuv2rgb.fs.h: yuv2rgb.fs
 	cd $(abs_srcdir) && \
-	xxd -i yuv2rgb.fs | sed s/}\;/,0x00}\;/ > $(abs_builddir)/yuv2rgb.fs.h
+	xxdi.pl yuv2rgb.fs | sed s/}\;/0x00}\;/ > $(abs_builddir)/yuv2rgb.fs.h
 
 yuv2rgb.vs.h: yuv2rgb.vs
 	cd $(abs_srcdir) && \
-	xxd -i yuv2rgb.vs | sed s/}\;/,0x00}\;/ > $(abs_builddir)/yuv2rgb.vs.h
+	xxdi.pl yuv2rgb.vs | sed s/}\;/0x00}\;/ > $(abs_builddir)/yuv2rgb.vs.h
 
 #because make bundle serahc in this dir
 install-data-local:
