diff -ur a/lemon.c b/lemon.c
--- a/lemon.c	2010-03-23 16:06:21.000000000 +0300
+++ b/lemon.c	2019-10-19 15:13:18.572603000 +0300
@@ -3109,6 +3109,20 @@
   }else{
     tpltname = pathsearch(lemp->argv0,templatename,0);
   }
+  /*Patch for gentoo so we don't have to add this directory to the path.*/
+  FILE *tpltnametest;
+  FILE *uniquelettercombo;
+  tpltnametest = fopen(tpltname,"rb");
+  if( tpltnametest==0 ){
+    uniquelettercombo = fopen("/usr/share/lemon/lempar.c","rb");
+    if (uniquelettercombo != NULL){
+      tpltname = "/usr/share/lemon/lempar.c";
+      fclose(uniquelettercombo);
+    }
+  }else{
+    fclose(tpltnametest);
+  }
+  /*End of Patch*/
   if( tpltname==0 ){
     fprintf(stderr,"Can't find the parser driver template file \"%s\".\n",
     templatename);
