diff -ur NPRQuake-SDL.orig/NPRQuakeSrc/common.c NPRQuake-SDL/NPRQuakeSrc/common.c
--- NPRQuake-SDL.orig/NPRQuakeSrc/common.c	2003-09-28 20:09:29.140832648 -0400
+++ NPRQuake-SDL/NPRQuakeSrc/common.c	2003-09-28 20:09:29.525774128 -0400
@@ -1775,11 +1775,16 @@
 // start up with GAMENAME by default (id1)
 //
 	COM_AddGameDirectory (va("%s/"GAMENAME, basedir) );
+	COM_AddGameDirectory (GENTOO_DATADIR"/"GAMENAME);
 
-	if (COM_CheckParm ("-rogue"))
+	if (COM_CheckParm ("-rogue")) {
 		COM_AddGameDirectory (va("%s/rogue", basedir) );
-	if (COM_CheckParm ("-hipnotic"))
+		COM_AddGameDirectory (GENTOO_DATADIR"/rogue");
+	}
+	if (COM_CheckParm ("-hipnotic")) {
 		COM_AddGameDirectory (va("%s/hipnotic", basedir) );
+		COM_AddGameDirectory (GENTOO_DATADIR"/hipnotic");
+	}
 
 //
 // -game <gamedir>
@@ -1790,6 +1795,7 @@
 	{
 		com_modified = true;
 		COM_AddGameDirectory (va("%s/%s", basedir, com_argv[i+1]));
+		COM_AddGameDirectory (va(GENTOO_DATADIR"/%s", com_argv[i+1]));
 	}
 
 //
