--- tclgpg.tcl	2012-02-11 23:14:06.000000000 +0400
+++ tclgpg.tcl	2012-02-11 23:25:22.000000000 +0400
@@ -1455,16 +1455,6 @@
                 FinishWithError $channels $commands "Decryption failed"
                 return
             }
-            KEYEXPIRED {
-                switch -- $operation {
-                    "" -
-                    verify {}
-                    default {
-                        FinishWithError $channels $commands "Key expired"
-                        return
-                    }
-                }
-            }
             KEYREVOKED {
                 switch -- $operation {
                     "" -
