# By L29Ah. Have fun!
diff -urN jumpnbump-1.50.old/main.c jumpnbump-1.50/main.c
--- jumpnbump-1.50.old/main.c	2009-10-12 02:18:27.531533704 +0400
+++ jumpnbump-1.50/main.c	2009-10-12 02:18:30.770698820 +0400
@@ -261,11 +261,11 @@
 
 typedef struct
 {
-	unsigned long cmd;
-	long arg;
-	long arg2;
-	long arg3;
-	long arg4;
+	uint32_t cmd;
+	int32_t arg;
+	int32_t arg2;
+	int32_t arg3;
+	int32_t arg4;
 } NetPacket;
 
 #define NETPKTBUFSIZE (4 + 4 + 4 + 4 + 4)
