--- src/ec_log.c	2007-08-17 07:46:06.000000000 -0500
+++ src/ec_log.c	2007-08-17 07:46:06.000000000 -0500
@@ -190,7 +190,7 @@
       if (fd->cfd == NULL)
          SEMIFATAL_ERROR("%s", gzerror(fd->cfd, &zerr));
    } else {
-      fd->fd = open(filename, O_CREAT | O_TRUNC | O_RDWR | O_BINARY);
+      fd->fd = open(filename, O_CREAT | O_TRUNC | O_RDWR | O_BINARY, 0644);
       if (fd->fd == -1)
          SEMIFATAL_ERROR("Can't create %s: %s", filename, strerror(errno));
    }
