--- configure.in
+++ configure.in
@@ -236,4 +236,11 @@ AC_SUBST(group)
 
 AC_CHECK_HEADERS(stdarg.h)
 
+AC_MSG_CHECKING(whether libc defines struct in6_pktinfo)
+AC_TRY_COMPILE([#include <netinet/in.h>],
+	[struct in6_pktinfo p6;],
+	[AC_MSG_RESULT(yes) 
+ CFLAGS="$CFLAGS -DHAVE_IN6_PKTINFO"],
+	[AC_MSG_RESULT(no)])
+
 AC_OUTPUT(Makefile)
diff --git wide/dhcp6.h wide/dhcp6.h
index 6afb517..5280a46 100644
--- dhcp6.h
+++ dhcp6.h
@@ -108,6 +108,13 @@ typedef uint64_t u_int64_t;
 #define DHCP6_IRT_DEFAULT 86400	/* 1 day */
 #define DHCP6_IRT_MINIMUM 600
 
+#ifndef HAVE_IN6_PKTINFO
+struct in6_pktinfo {
+	struct in6_addr ipi6_addr; /* src/dst IPv6 address */
+	unsigned int ipi6_ifindex; /* send/recv interface index */
+};
+#endif /* HAVE_IN6_PKTINFO */
+
 /* DUID: DHCP unique Identifier */
 struct duid {
 size_t duid_len;	/* length */
