diff -up mc-4.6.2-pre1/configure.extensions mc-4.6.2-pre1/configure
--- mc-4.6.2-pre1/configure.extensions	2007-09-11 16:33:32.000000000 +0200
+++ mc-4.6.2-pre1/configure	2008-03-07 14:28:51.000000000 +0100
@@ -5736,7 +5736,7 @@
 fi
 
 
-for ac_prog in gnome-moz-remote mozilla konqueror opera netscape
+for ac_prog in firefox gnome-moz-remote mozilla konqueror opera netscape
 do
   # Extract the first word of "$ac_prog", so it can be a program name with args.
 set dummy $ac_prog; ac_word=$2
diff -up mc-4.6.2-pre1/lib/mc.ext.in.extensions mc-4.6.2-pre1/lib/mc.ext.in
--- mc-4.6.2-pre1/lib/mc.ext.in.extensions	2006-12-28 04:57:01.000000000 +0100
+++ mc-4.6.2-pre1/lib/mc.ext.in	2008-03-07 14:28:31.000000000 +0100
@@ -119,6 +119,11 @@
 	Open=%cd %p#utar
 	View=%view{ascii} bzip2 -dc %f 2>/dev/null | tar tvvf -
 
+# .tar.lzma, .tlz
+regex/\.t(ar\.lzma|lz)$
+	Open=%cd %p#utar
+	View=%view{ascii} lzma -dc %f 2>/dev/null | tar tvvf -
+
 # .tar.F - used in QNX
 regex/\.tar\.F$
 	# Open=%cd %p#utar
@@ -220,14 +225,25 @@
 
 # C
 shell/.c
-	Open=%var{EDITOR:vi} %f
+	Include=editor
 
 # Fortran
 shell/.f
-	Open=%var{EDITOR:vi} %f
+	Include=editor
 	
 # Header
-regex/\.(h|hpp)$
+regex/\.([Hh]|[Hh]pp|HPP)$
+	Include=editor
+
+# Asm
+regex/\.([Ss]|[Aa]sm|ASM)$
+	Include=editor
+
+# C++
+regex/\.(C|cc|[Cc]pp|CPP)$
+	Include=editor
+
+include/editor
 	Open=%var{EDITOR:vi} %f
 
 # Object
@@ -251,10 +267,12 @@
 
 # GNU Info page
 type/^Info\ text
-	Open=info -f %f
+	#Open=info -f %f
+	Open=pinfo %f
 
 shell/.info
-	Open=info -f %f
+	#Open=info -f %f
+	Open=pinfo %f
 
 # Manual page
 # Exception - .so libraries are not manual pages
@@ -298,6 +316,10 @@
 	Open=case %d/%f in */log/*|*/logs/*) bzip2 -dc %f ;; *) bzip2 -dc %f | nroff @MAN_FLAGS@ @MANDOC@ ;; esac | %var{PAGER:more}
 	View=%view{ascii,nroff} case %d/%f in */log/*|*/logs/*) bzip2 -dc %f ;; *) bzip2 -dc %f | nroff @MAN_FLAGS@ @MANDOC@ ;; esac
 
+regex/([^0-9]|^[^\.]*)\.([1-9][A-Za-z]*|[ln])\.lzma$
+	Open=case %d/%f in */log/*|*/logs/*) lzma -dc %f ;; *) lzma -dc %f | nroff @MAN_FLAGS@ @MANDOC@ ;; esac | %var{PAGER:more}
+	View=%view{ascii,nroff} case %d/%f in */log/*|*/logs/*) lzma -dc %f ;; *) lzma -dc %f | nroff @MAN_FLAGS@ @MANDOC@ ;; esac
+
 
 ### Images ###
 
@@ -328,18 +350,24 @@
 type/^Netpbm
 	Include=image
 
+type/^JNG
+	Include=image
+
+type/^MNG
+	Include=image
+
 shell/.xcf
 	Open=(gimp %f &)
 
 shell/.xbm
-	Open=bitmap %f
+	Include=image
 
 shell/.xpm
 	Include=image
 	View=sxpm %f
 
 include/image
-	Open=if [ "$DISPLAY" = "" ]; then zgv %f; else (gqview %f &); fi
+	Open=if [ "$DISPLAY" = "" ]; then zgv %f; else (xdg-open %f &); fi
 	View=%view{ascii} identify %f
 	#View=%view{ascii} asciiview %f
 
@@ -347,7 +375,8 @@
 ### Sound files ###
 
 regex/\.([wW][aA][vV]|[sS][nN][dD]|[vV][oO][cC]|[aA][uU]|[sS][mM][pP]|[aA][iI][fF][fF]|[sS][nN][dD])$
-       Open=if [ "$DISPLAY" = "" ]; then play %f; else (xmms %f >/dev/null 2>&1 &); fi
+	Open=play %f
+#       Open=if [ "$DISPLAY" = "" ]; then play %f; else (xmms %f >/dev/null 2>&1 &); fi
 
 regex/\.([mM][oO][dD]|[sS]3[mM]|[xX][mM]|[iI][tT]|[mM][tT][mM]|669|[sS][tT][mM]|[uU][lL][tT]|[fF][aA][rR])$
        Open=mikmod %f
@@ -357,11 +386,16 @@
        Open=vplay -s 22 %f
 
 regex/\.([mM][pP]3)$
-	Open=if [ "$DISPLAY" = "" ]; then mpg123 %f; else (xmms %f >/dev/null 2>&1 &); fi
-	View=%view{ascii} mpg123 -vtn1 %f 2>&1 | sed -n '/^Title/,/^Comment/p;/^MPEG/,/^Audio/p'
+	Include=audio
+#	Open=if [ "$DISPLAY" = "" ]; then mpg123 %f; else (xmms %f >/dev/null 2>&1 &); fi
+#	View=%view{ascii} mpg123 -vtn1 %f 2>&1 | sed -n '/^Title/,/^Comment/p;/^MPEG/,/^Audio/p'
+
+regex/\.([mM][kK][aA])$
+	Include=audio
 
 regex/\.([oO][gG][gG])$
-	Open=if [ "$DISPLAY" = "" ]; then ogg123 %f; else (xmms %f >/dev/null 2>&1 &); fi
+	Open=ogg123 %f
+#	Open=if [ "$DISPLAY" = "" ]; then ogg123 %f; else (xmms %f >/dev/null 2>&1 &); fi
 	View=%view{ascii} ogginfo %s
 
 regex/\.([mM][iI][dD][iI]?|[rR][mM][iI][dD]?)$
@@ -371,11 +405,15 @@
 	Open=mplayer -vo null %f
 	View=%view{ascii} mplayer -quiet -slave -frames 0 -vo null -ao null -identify %f 2>/dev/null | tail +13 || file %f
 
+include/audio
+	Open=mplayer %f
+	View=%view{ascii} mplayer -identify -vo null -ao null -frames 0 %f 2>&1 | sed -n '/^ID_/p'
 
 ### Play lists ###
 
 regex/\.([mM]3[uU]|[pP][lL][sS])$
-	Open=if [ -z "$DISPLAY" ]; then mplayer -vo null -playlist %f; else (xmms -p %f >/dev/null 2>&1 &); fi
+	Open=mplayer -vo null -playlist %f
+#	Open=if [ -z "$DISPLAY" ]; then mplayer -vo null -playlist %f; else (xmms -p %f >/dev/null 2>&1 &); fi
 
 
 ### Video ###
@@ -395,20 +433,27 @@
 regex/\.([mM][pP]4|[mM][pP][eE]?[gG])$
 	Include=video
 
+regex/\.([mM][kK][vV])$
+	Include=video
+
 regex/\.([vV][oO][bB])$
 	Include=video
 
 regex/\.([wW][mM][vV])$
 	Include=video
 
+regex/\.([fF][lL][iIcC])$
+	Include=video
+
 regex/\.([oO][gG][mM])$
 	Include=video
 
 regex/\.([rR][aA]?[mM])$
-	Open=(realplay %f >/dev/null 2>&1 &)
+	Include=video
 
 include/video
 	Open=(mplayer %f >/dev/null 2>&1 &)
+	View=%view{ascii} mplayer -identify -vo null -ao null -frames 0 %f 2>&1 | sed -n '/^ID_/p'
 	#Open=(gtv %f >/dev/null 2>&1 &)
 	#Open=(xanim %f >/dev/null 2>&1 &)
 
@@ -417,12 +462,13 @@
 
 # Postscript
 type/^PostScript
-	Open=(gv %f &)
+	Open=(evince %f >/dev/null 2>&1 &)
 	View=%view{ascii} ps2ascii %f
 
 # PDF
 type/^PDF
-	Open=(xpdf %f &)
+	Open=(xdg-open %f >/dev/null 2>&1 &)
+	#Open=(xpdf %f >/dev/null 2>&1 &)
 	#Open=(acroread %f &)
 	#Open=(ghostview %f &)
 	View=%view{ascii} pdftotext %f -
@@ -433,7 +479,7 @@
 # html
 regex/\.([hH][tT][mM][lL]?)$
 	Open=(if test -n "@X11_WWW@" && test -n "$DISPLAY"; then (@X11_WWW@ file://%d/%p &) 1>&2; else links %f || lynx -force_html %f || ${PAGER:-more} %f; fi) 2>/dev/null
-	View=%view{ascii} lynx -dump -force_html %f
+	View=%view{ascii} links -dump %f
 
 # StarOffice 5.2
 shell/.sdw
@@ -450,22 +496,27 @@
 
 # Microsoft Word Document
 regex/\.([Dd][oO][cCtT]|[Ww][rR][iI])$
-	Open=(abiword %f >/dev/null 2>&1 &)
+	Open=(ooffice %f &)
+#	Open=(abiword %f >/dev/null 2>&1 &)
 	View=%view{ascii} catdoc -w %f || word2x -f text %f - || strings %f
 type/^Microsoft\ Word
-	Open=(abiword %f >/dev/null 2>&1 &)
+	Open=(ooffice %f &)
+#	Open=(abiword %f >/dev/null 2>&1 &)
 	View=%view{ascii} catdoc -w %f || word2x -f text %f - || strings %f
 
 # RTF document
 regex/\.([rR][tT][fF])$
-	Open=(abiword %f >/dev/null 2>&1 &)
+	Open=(ooffice %f &)
+#	Open=(abiword %f >/dev/null 2>&1 &)
 
 # Microsoft Excel Worksheet
 regex/\.([xX][lL][sSwW])$
-	Open=(gnumeric %f >/dev/null 2>&1 &)
+	Open=(ooffice %f &)
+#	Open=(gnumeric %f >/dev/null 2>&1 &)
 	View=%view{ascii} xls2csv %f || strings %f
 type/^Microsoft\ Excel
-	Open=(gnumeric %f >/dev/null 2>&1 &)
+	Open=(ooffice %f &)
+#	Open=(gnumeric %f >/dev/null 2>&1 &)
 	View=%view{ascii} xls2csv %f || strings %f
 
 # Use OpenOffice.org to open any MS Office documents
@@ -542,6 +593,11 @@
 	Open=gzip -dc %f | %var{PAGER:more}
 	View=%view{ascii} gzip -dc %f 2>/dev/null
 
+# lzma
+regex/\.lzma$
+	Open=lzma -dc %f | %var{PAGER:more}
+	View=%view{ascii} lzma -dc %f 2>/dev/null
+
 
 ### Default ###
 
diff -up mc-4.6.2-pre1/configure.ac.extensions mc-4.6.2-pre1/configure.ac
--- mc-4.6.2-pre1/configure.ac.extensions	2007-09-10 16:25:30.000000000 +0200
+++ mc-4.6.2-pre1/configure.ac	2008-03-07 14:28:51.000000000 +0100
@@ -83,7 +83,7 @@
 AC_CHECK_TOOL(AR, ar, ar)
 
 dnl Only list browsers here that can be run in background (i.e. with `&')
-AC_CHECK_PROGS(X11_WWW, [gnome-moz-remote mozilla konqueror opera netscape])
+AC_CHECK_PROGS(X11_WWW, [firefox gnome-moz-remote mozilla konqueror opera netscape])
 
 dnl
 dnl Ovverriding mmap support.  This has to be before AC_FUNC_MMAP is used.
