diff -ruN aegisub-old/configure.in aegisub-9999/configure.in
--- aegisub-old/configure.in	2009-05-15 01:01:40.065849105 +0400
+++ aegisub-9999/configure.in	2009-05-15 01:05:22.638598331 +0400
@@ -520,9 +520,7 @@
   AC_AGI_COMPILE([PortAudio], [portaudio], [$PORTAUDIO_CFLAGS], [$PORTAUDIO_LDFLAGS],[
 #include <portaudio.h>
 int main(void) {
-  PaError err = Pa_Initialize();
-  if (err != paNoError) return 1;
-  return 0;
+  return 1;
 } ])
 fi
 
@@ -555,9 +553,7 @@
   AC_AGI_COMPILE([PuleseAudio], [pulseaudio], [$LIBPULSE_CFLAGS], [$LIBPULSE_LIBS],[
 #include <pulse/pulseaudio.h>
 int main(void) {
-pa_threaded_mainloop *mainloop = pa_threaded_mainloop_new();
-if (!mainloop) return 1;
-return 0;
+return 1;
 } ])
 fi
 
@@ -593,11 +589,7 @@
 #include <AL/alc.h>
 #endif
 int main(void) {
-  ALCdevice *device = alcOpenDevice(0);
-  if (!device) return 1;
-  ALCcontext *context = alcCreateContext(device, 0);
-  if (!context) return 1;
-  return 0;
+  return 1;
 } ])
 fi
 
