diff -ru plan9/bin/9a plan9.new/bin/9a
--- plan9/bin/9a	2010-03-17 22:25:01.000000000 +0000
+++ plan9.new/bin/9a	2011-03-21 03:20:27.000000000 +0000
@@ -1,15 +1,21 @@
 #!/bin/sh
 
-if [ $# != 1 ]; then
-	echo 'usage: 9a file.s' 1>&2
-	exit 1
-fi
-
 aflags=""
 case "`uname`" in
 Darwin)
 	aflags="-arch `arch`"
 esac
 
-out=`echo $1 | sed 's/\.s$//;s/$/.o/'`
-exec as $aflags -o $out $1
+for arg; do
+	case $arg in
+	-*)
+		opts="$opts $arg"
+		;;
+	*.s)
+		src=$arg
+		dst=$(echo $arg | sed 's/\.s/\.o/')
+		;;
+	esac
+done
+
+exec as $aflags $opts -o $dst $src
diff -ru plan9/src/mkhdr plan9.new/src/mkhdr
--- plan9/src/mkhdr	2010-04-10 03:59:41.000000000 +0000
+++ plan9.new/src/mkhdr	2011-03-20 23:57:26.000000000 +0000
@@ -29,7 +29,7 @@
 INSTALL=install
 CFLAGS=
 LDFLAGS=
-AFLAGS=
+AFLAGS=--noexecstack
 CLEANFILES=
 NUKEFILES=
 YACC=9 yacc
