diff --git a/dataobj/route.cc b/dataobj/route.cc
index d5ef2e6..6b2a1a0 100644
--- a/dataobj/route.cc
+++ b/dataobj/route.cc
@@ -8,11 +8,6 @@
 #include <stdio.h>
 #include <string.h>
 
-#include <limits.h>
-#ifdef UINT_MAX
-#define MAXUINT32 UINT_MAX
-#endif
-
 #include "../simworld.h"
 #include "../simcity.h"
 #include "../simintr.h"
diff --git a/simworld.h b/simworld.h
index 547a18d..bae403e 100644
--- a/simworld.h
+++ b/simworld.h
@@ -43,7 +43,14 @@
 #define snprintf sprintf_s
 #else
 #define sprintf_s snprintf
-#endif 
+#endif
+
+#include <limits.h>
+#ifdef UINT_MAX
+#define MAXUINT32 UINT_MAX
+#else
+# define MAXUINT32 4294967295
+#endif
 
 struct event_t;
 struct sound_info;
