--- makeobj/Makefile.old 2013-07-19 20:00:57.738576706 +0400
+++ makeobj/Makefile 2013-07-19 20:00:57.738576706 +0400
@@ -121,5 +121,12 @@
 SHARED_SOURCES += ../dataobj/freelist.cc
 SHARED_SOURCES += ../dataobj/tabfile.cc
+SHARED_SOURCES += ../dataobj/loadsave.cc
+SHARED_SOURCES += ../dataobj/umgebung.cc
+SHARED_SOURCES += ../dataobj/einstellungen.cc
+SHARED_SOURCES += ../dataobj/network.cc
+SHARED_SOURCES += ../gui/messagebox.cc
 SHARED_SOURCES += ../simdebug.cc
 SHARED_SOURCES += ../simmem.cc
+SHARED_SOURCES += ../simgraph$(COLOUR_DEPTH).cc
+SHARED_SOURCES += ../simwin.cc
+SHARED_SOURCES += ../simskin.cc
 SHARED_SOURCES += ../utils/simstring.cc
