--- SConstruct	2008-09-13 05:37:59.000000000 +0400
+++ SConstruct	2009-05-17 18:31:48.000000000 +0400
@@ -10,8 +10,8 @@
 env = Environment()
 
 env['CC'] = 'gcc'
-env['LINKFLAGS'] = Split('-pipe -Wall -Werror')
-env['CCFLAGS'] = Split('-pipe -Wall -Werror -std=c99 -Wno-switch -Wno-unused -Wno-missing-braces -Wno-parentheses -Wno-uninitialized -fno-strict-aliasing -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DTEXT_DOMAIN=\\"zfs-fuse\\" -DLINUX_AIO')
+env['LINKFLAGS'] = Split('-pipe -Wall')
+env['CCFLAGS'] = Split('-pipe -Wall -std=c99 -Wno-switch -Wno-unused -Wno-missing-braces -Wno-parentheses -Wno-uninitialized -fno-strict-aliasing -D_GNU_SOURCE -D_FILE_OFFSET_BITS=64 -D_REENTRANT -DTEXT_DOMAIN=\\"zfs-fuse\\" -DLINUX_AIO')
 
 debug = int(ARGUMENTS.get('debug', '1'))
 
