--- synaptics-0.14.6.orig/synaptics.c	2008-01-18 23:32:59.000000000 -0500
+++ synaptics-0.14.6/synaptics.c	2008-01-18 23:32:15.000000000 -0500
@@ -617,10 +617,10 @@
 			    miPointerGetMotionBufferSize());
 
     /* X valuator */
-    xf86InitValuatorAxisStruct(dev, 0, 0, -1, 1, 0, 1);
+    xf86InitValuatorAxisStruct(dev, 0, -1, -1, 1, 0, 1);
     xf86InitValuatorDefaults(dev, 0);
     /* Y valuator */
-    xf86InitValuatorAxisStruct(dev, 1, 0, -1, 1, 0, 1);
+    xf86InitValuatorAxisStruct(dev, 1, -1, -1, 1, 0, 1);
     xf86InitValuatorDefaults(dev, 1);
 
     xf86MotionHistoryAllocate(local);

