slocate-2.x databases are incompat with slocate-3.x

--- src/slocate.c
+++ src/slocate.c
@@ -435,7 +435,7 @@
 		 * cancelling the + */
 		len = path_len+code_num;
 		if (len <= 0) {
-			if (!report_error(g_data, FATAL, "set_path_head: path_head len <= 0: %d\n", len))
+			if (!report_error(g_data, FATAL, "set_path_head: path_head len <= 0: %d\nPerhaps your database is still the old format and you need to run `updatedb` ?\n", len))
 			    goto EXIT;
 		}
 		if (!(path_head = realloc(path_head, (sizeof(char) * (len+1))))) {
