diff -ruN naev-0.3.8.orig/build/Makefile.common naev-0.3.8/build/Makefile.common
--- naev-0.3.8.orig/build/Makefile.common	2009-04-19 16:20:37.260362651 +0400
+++ naev-0.3.8/build/Makefile.common	2009-04-19 16:27:41.966070370 +0400
@@ -50,5 +50,5 @@
 						scripts/*/*.lua \
 						scripts/*/*/*.lua )
 DATA_NEWS :=	$(wildcard dat/news/*.lua)
-DATA      := 	ndata
+DATA      := 	${GAMES_DATADIR}/${PN}/ndata
 DATAFILES := 	$(DATA_MISC) $(DATA_AI) $(DATA_GFX) $(DATA_XML) $(DATA_SND) $(DATA_MISN) $(DATA_SCRIPTS) $(DATA_NEWS)
