diff --git a/Makefile b/Makefile
index 57be0cd..7ebcf3f 100644
--- a/Makefile
+++ b/Makefile
@@ -12,8 +12,8 @@ SRC = src
 CXXFLAGS = -I$(SRC) -std=c++11 -MD $(OPTDEBUGFLAGS) -Wall -Wshadow -Wsign-compare -Werror
 LIBS = -lm
 
-DESTDIR = /usr/local
-ICEBOX = /usr/local/share/icebox
+DESTDIR = /usr
+ICEBOX = /usr/share/icebox
 
 .PHONY: all
 all: bin/arachne-pnr share/arachne-pnr/chipdb-1k.bin share/arachne-pnr/chipdb-8k.bin
diff --git a/tests/simple/run-test.sh b/tests/simple/run-test.sh
index 7a455d7..b0db888 100644
--- a/tests/simple/run-test.sh
+++ b/tests/simple/run-test.sh
@@ -4,7 +4,7 @@ set -ex
 
 arachne_pnr="../../bin/arachne-pnr"
 devices='1k 8k'
-: ${ICEBOX:=/usr/local/share/icebox}
+: ${ICEBOX:=/usr/share/icebox}
 
 rm -f txt.sum
 
